### PR TITLE
ZMP modification

### PR DESCRIPTION
**Pull request by ymshi449**
1. Fixed accumulation bugs
2. Fixed v0 Fock matrix bugs when the input is not HF/DFT by using natural orbitals
3. Modified interface of on_grid_esp. User should not modify the wfn.Da or wfn.Db.

#TODO
Added error cancellation.

### CODE REVIEW

File: n2v/grid/grider.py [hunk 1]
```diff
@@ -455,6 +465,12 @@ def on_grid_esp(self, grid=None, vpot=None, wfn=None):
         hartree = - 1.0 * (vext + esp)
         v_fa = (1 - 1.0 / (self.nalpha + self.nbeta)) * hartree
 
+        if Da is not None:
+            self.wfn.Da().np[:] = Da_temp
+            print(np.linalg.norm(Da_temp))
```
Comment (VHchavez) [inline n2v/grid/grider.py:470]: Why are we printing the norm? Was this used for testing?

File: n2v/inverter.py [hunk 2]
```diff
@@ -147,7 +149,8 @@ def generate_jk(self, gen_K=False, memory=2.50e9):
         jk.set_memory(int(memory)) 
         jk.set_do_K(gen_K)
         jk.initialize()
-        
+        jk.initialize()
```
Comment (VHchavez) [inline n2v/inverter.py:152]: Is this duplication a typo?

File: n2v/inverter.py [hunk 3]
```diff
@@ -384,7 +387,25 @@ def generate_components(self, guide_potential_components):
         self.vb = np.zeros_like(self.T)
 
         N = self.nalpha + self.nbeta
-        self.J0, _ = self.form_jk(self.ct[0], self.ct[1])
+
+        if self.J0 is None:
```
Comment (VHchavez) [inline n2v/inverter.py:391]: I know you mentioned the use for natural orbitals, but can you direct me to some references that you used to understand it better?

File: n2v/methods/zmp.py [hunk 4]
```diff
@@ -248,22 +252,34 @@ def zmp_scf(self,
...
-            if np.abs(grid_diff_old) < np.abs(grid_diff):
-                print("\nZMP halted. Density Difference is unable to be reduced")
-                break
-
-            grid_diff_old = grid_diff
-            print(f"SCF Converged for lambda:{int(lam_i):5d}. Max density difference: {grid_diff}")
-
-            self.Da = Da
-            self.Db = Db
+            # Why do we need this? Basis set -> grid is expensive.
```
Comment (VHchavez) [inline n2v/methods/zmp.py:255]: I know there is a relationship between the accuracy of P_nm and n(r), but they're not identical. I'd still find value in having the accuracy on the grid. Maybe just outside scf?

File: n2v/methods/zmp.py [hunk 5]
```diff
@@ -248,22 +252,34 @@ def zmp_scf(self,
...
+            # grid_diff = np.max(np.abs(D0 - density_current))
+            # if np.abs(grid_diff_old) < np.abs(grid_diff):
+            #     print("\nZMP halted. Density Difference is unable to be reduced")
+            #     break
+            #
+            # grid_diff_old = grid_diff
+            print(f"SCF Converged for lambda:{int(lam_i):5d}. Max density difference: {np.linalg.norm(Da - self.nt[0])}")
+
+            #VXC is hartree-like Potential. We remove Fermi_Amaldi Guess.
+            self.proto_density_a += lam_i * (Da - self.nt[0]) * self.mixing
```
Comment (VHchavez) [inline n2v/methods/zmp.py:266]: I know the mixing here is reducing the amount of the addition to the accumulation, but shouldn't we be also removing from self.proto_density_a as well? They're similar ideas, but not the same.

**Review by VHchavez — COMMENTED** [review]
.

File: n2v/grid/grider.py (hunk 1 above)
Comment (ymshi449) [reply n2v/grid/grider.py:470]: Yes I will delete that.

File: n2v/inverter.py (hunk 2 above)
Comment (ymshi449) [reply n2v/inverter.py:152]: Yes.

File: n2v/inverter.py (hunk 3 above)
Comment (ymshi449) [reply n2v/inverter.py:391]: I don't have a paper for this and it is strictly not natural orbitals. The main idea is to find effective orbitals and effective (fractional) occupations that will reproduce the correct density matrices.

File: n2v/methods/zmp.py (hunk 4 above)
Comment (ymshi449) [reply n2v/methods/zmp.py:255]: Agree. But I am not sure if we want to break when the density error stops updating.

File: n2v/methods/zmp.py (hunk 5 above)
Comment (ymshi449) [reply n2v/methods/zmp.py:266]: Could you say it again? I don't get it.

Comment (VHchavez) [reply n2v/methods/zmp.py:268]: You add a portion of the new potential. But you're not removing anything from the old potential as in (1-mixing) * vc_old + mixing * vc_new
Just trying to understand if your intention was to do a mix, or just to limit how much of the new potential gets added.

Comment (ymshi449) [reply n2v/methods/zmp.py:270]: Yes you are right. This is strictly speaking not a liner mixing, but rather an accumulation coefficient. The thing is linear mixing does not work smoothly with accumulation. I added lines (commented off) for real linear mixing that you can try.

**Comment by VHchavez** [comment]
@ymshi449 
lgtm! If it's all ready, let's merge it. Would love to play with it.

**Review by VHchavez — APPROVED** [review]

**Pull request merged**